### ISSUE TITLE
Using Private registries with basic authentication 

### ISSUE THREAD

**Issue by galal-hussein** [rancher/rancher#13339]
**Rancher versions:**
rancher/rancher:master

Currently using private registry with basic authentication will succeed for deploying k8s components but will fail for deploying the pods because kubelet has no idea what credentials are being used and it fails to pull the pause image and pods images.

**Comment by apricote** [comment]
We are also experiencing this issue. We tried this either with the beta or some RC and as far as I can remember it was working back then.

Current workaround we are using is to edit the YAML of the Deployment/Workload and add the `imagePullSecrets` manually:

```yaml
kind: Deployment
name: workload-name
spec:
  template:
    spec:
      containers:
      - image: ...
      imagePullSecrets:
      - name: registry-credentials-name
```

While this is no maintainable solution, it allows us to experiment with Rancher 2.

**Comment by strowi** [comment]
Still an issue with 2.0.5.
This is currently a blocker for migrating our prod-workload to rancher2.
ImagePullSecrets 'right now' would be too much overhead for multiple teams with multiple projects..

**Comment by Terminator637** [comment]
This is really bug. Please, fix it.

**Comment by gillarda** [comment]
Having the same issue here. Eagerly waiting for a fix.

**Comment by shuckepco** [comment]
Still present in 2.0.6 - really a huge blocker. We are using an AWS ECR registry. In contrast to #13332, it does not work when deployed via UI either.

However, working around the issue by manually editing the YAML definition for each deployment is tolerable as long as we are not in production...

**Comment by killfill** [comment]
Hi, just my 0.1cents here...

Not sure if i would consider this as a bug really.

It looks like previous rancher did insert the secret by itself:

The good: user's didn't need to specify any secret to use private repo

The bad: Magic (i don't like magic too much). User's did not really learned how to use secerts and could not use 2 different private repos.

If you define the regitry secret in rancher ui, and then reference that secret in the deployment, things works well.

:-D

**Comment by shuckepco** [comment]
It's a bit of a philosophical question... Is it a bug or a missing sub-feature? IMHO I would call it a bug:
- Rancher 2.x allows the creation of private registries that need basic auth.
- You are able to deploy workloads via UI using these registries.
- You get an error that you need to authenticate when you try to run it.
- There is no way to resolve this issue, only a workaround by manually correcting erroneous settings.

@alena1108 explained the differences between 1.6 and 2.x in #13332. That is entirely clear to me. But that is not the point. (Don't like too much "magic" either :-) ... But I like simplicity and efficiency of everyday tasks) The point is that the "magic" is missing in this _specific_ case and it is existent in others (e.g. see #13676). Furthermore, you cannot choose a less "magical" way and select the private repo to pull from when deploying workloads via Rancher UI. You need to create a non-functioning workload first in order to fix it by manually editing a YAML file.

Nevertheless, it brings up the question on how to resolve this.
1. Make the registry selectable when deploying workloads (drop-down menu, searchable input field...)
2. Auto-detect the registry by using the Docker image URI
3. ...?

**Comment by strowi** [comment]
Another use-case for the 1.6 beahviour:
Unless i'm missing a step..

We are moving repos to gitlab, which allows per-repo authentication also for the docker-registry.
With 1.6 as admin i could just provide a global account, that could pull every image on k8s.
With 2.0 every project would need it's own imagePullSecret - which one would you choose? 
Also a global one -> authentication would be useless, everyone knows the global secret?
Per Repo? Also very unconvenient, also everyone with access could just grab my token...?

**Comment by deniseschannon** [comment]
@galal-hussein is the failure on the local cluster or any k8s cluster that gets added into rancher?

**Comment by alena1108** [comment]
@deniseschannon the issue would be on all the clusters. @galal-hussein we should address the original issue first - addons not being deployed correctly. For that, in airgapped setup on rke up you'd have to create a secret with cluster private repo credentials, and reference the secret in the generated addon templates

**Comment by alena1108** [comment]
The RKE issue is tracked separately: https://github.com/rancher/rke/issues/588

**Comment by strowi** [comment]
As far as i understand rancher2 uses rke underneath and it is possible to add rke-options into the rancher2 cluster-yaml-editor? 
If so, it would be interesting which version of rke rancher2 uses and if it is possible to add private registries like the doc says: https://rancher.com/docs/rke/v0.1.x/en/config-options/private-registries/

**Comment by mgresser** [comment]
I attempted this and it started to try using the private registry for all the system images. It also refuses to allow me to remove the setting which leaves my cluster in a bad state as it's now trying to install itself as if it was air gapped. I tried adding the appropriate images to my private registry with mixed results. Some hosts were able to pull but others seem unable to.

**Comment by shubb30** [comment]
There is also a problem with the workaround of adding the `imagePullSecrets` to the YAML.  If you ever Edit the workload through the web UI, it rewrites the whole YAML and removes the `imagePullSecrets`, so it breaks again.  You will need to edit the Kubernetes YAML every time you want to bump the image version or make any changes to the workload.

**Comment by loganhz** [comment]
I guess some of you are talking about this issue https://github.com/rancher/rancher/issues/15994?

**Comment by alena1108** [comment]
The comments to this issue are talking about different things, so here is the recap:

* Issue 1 - RKE clusters fail to deploy addons in airgapped setup. Tracked in https://github.com/rancher/rke/issues/588 - resolved
* Issue 2 - Add UI support for passing private repo and credentials on per cluster basis (API support is already in). Tracked in https://github.com/rancher/rancher/issues/15554, targeted 2.2)

The credentials passed on RKE cluster create get used **only** for system rke components and addons.

Now to the the recommended workflow on private registries for users workload (discussed in  comments https://github.com/rancher/rancher/issues/13339#issuecomment-410728588, https://github.com/rancher/rancher/issues/13339#issuecomment-412273680,https://github.com/rancher/rancher/issues/13339#issuecomment-410714464 ). In Rancher UI, you have to define registry credential either on a project level, or on the namespace level:
![image](https://user-images.githubusercontent.com/1562176/47119616-66c87600-d220-11e8-9f0e-d6c1917ea2cc.png)

There is currently no way to define the cluster or global level private registry. And honestly saying, implementing it might go against k8s best practices where registry secret **always** limited to a namespace (https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/). 

Now there are known issues with secrets used in the workloads:

* https://github.com/rancher/rancher/issues/15994, same as https://github.com/rancher/rancher/issues/15725 // resolved, will be available in 2.2

**Comment by alena1108** [comment]
For all registry related issues that are not covered in https://github.com/rancher/rancher/issues/13339#issuecomment-430812349, please file a separate bug. Closing this issue
